gdlsc.go

```diff
@@ -1,104 +1,73 @@
-	Name string
+	License string
+	File    string
 }
 
-// uses github api to fetch license based on link generated
-func getLicense(dependency string, ch chan message) {
-	url := formatLink(dependency)
-	resp, err := http.Get(url)
-	if err != nil {
-		panic(err)
-	}
-
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		panic(err)
-	}
-
-	cleanBody := &response{}
-	if err := json.Unmarshal([]byte(body), &cleanBody); err != nil {
-		panic(err)
-	}
-
-	link := cleanBody.Html_URL
-	license := cleanBody.License.Name
-	if link == "" {
-		link = "Not Found"
-	}
-
-	ch <- message{
-		Dependency: dependency,
-		License:    license,
-		Link:       link,
+func getLicense(input chan string, output chan response) {
+	for file := range input {
+		l, err := license.NewFromFile(file)
+		splitFile := strings.Split(file, "/")
+		newFile := splitFile[1 : len(splitFile)-1]
+		file = strings.Join(newFile, "/")
+		if err != nil {
+			output <- response{File: file, License: err.Error()}
+		} else {
+			output <- response{File: file, License: l.Type}
+		}
 	}
 }
 
-// get just the owner and repo and generate github license link
-func formatLink(link string) string {
-	strippedURL := strings.Split(link, "/")
-	authorAndRepo := strings.Join(strippedURL[1:3], "/")
-	licenseURL := fmt.Sprintf(URL, authorAndRepo)
-	return licenseURL
-}
+func main() {
+	root := "vendor/"
+	i := 0
+	input := make(chan string)
+	output := make(chan response)
 
-// removes duplicate links
-func removeDuplicates(links []string) []string {
-	linksMap := make(map[string]int)
-	for _, link := range links {
-		linksMap[link] = 1
+	for a := 0; a < 5; a++ {
+		go getLicense(input, output)
 	}
-	linksSlice := make([]string, 0, len(linksMap))
-	for link := range linksMap {
-		linksSlice = append(linksSlice, link)
-	}
-	return linksSlice
-}
+	var files []string
+	filepath.Walk(root, func(path string, f os.FileInfo, err error) error {
+		if err != nil {
+			fmt.Println("No vendor folder found. Aborting..")
+			return err
+		}
+		if !f.IsDir() {
+			splitPath := strings.Split(path, "/")
+			rawPath := splitPath[len(splitPath)-1]
+			if strings.Contains(strings.ToUpper(rawPath), "LICENSE") {
+				files = append(files, path)
+			}
+		}
+		return nil
+	})
 
-func main() {
-	out, err := exec.Command("sh", "-c", `go list -f '{{ join .Imports "\n"}}' ./... | grep github`).Output()
-	if err != nil {
-		log.Fatal(err)
-	}
-	depSlice := strings.Split(string(out), "\n")
-	rawDeps := depSlice[:len(depSlice)-1] //remove trailing empty string
-	dependencies := removeDuplicates(rawDeps)
-	ch := make(chan message, len(dependencies)-1)
-	for _, rawDep := range rawDeps {
-		go getLicense(rawDep, ch)
+	if len(files) == 0 {
+		return
 	}
 
-	for i := 0; i < len(rawDeps); i++ {
-		result := <-ch
-		fmt.Printf("%v ===>: %v (%v)\n", result.Dependency, result.License, result.Link)
+	go func() {
+		for _, v := range files {
+			input <- v
+		}
+	}()
+	for out := range output {
```

Comment:
I think a nicer way to achieve the same is:

for i := 0; i < len(files); i++ {
  out := <- output
  ...
}